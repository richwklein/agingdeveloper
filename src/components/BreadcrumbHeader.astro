---
/**
 * Component used for rendering a breadcrumb at the top of the page.
 * If tail is null then just the name of the head page is shown.
 * Otherwise, a link to the head page and the name of the tail are shown.
 *
 * @param {Props} props - The head and tail parts of the breadcrumb.
 * @return {Function} an Astro component factory function.
 *
 * @example
 * <BreadcrumbBlock head={{name: "Archive", path="/article"}} tail="1" />
 */
import InternalLink from "./InternalLink.astro";

interface Props {
  head: {
    name: string;
    path: string;
  };
  tail?: string;
  class?: string;
}

const { head, tail, class: className } = Astro.props;
---

<header class:list={[className, "flex w-full items-center"]}>
  <span class:list={["mr-2 h-px flex-grow rounded bg-gray-300"]}></span>
  <ol class:list={["text-primary inline-flex items-center space-x-0 capitalize md:space-x-3"]}>
    {
      tail == null ? (
        <li>{head.name}</li>
      ) : (
        <li class:list={["after:mx-1 after:text-gray-300 after:content-['/']"]}>
          <InternalLink to={head.path} class:list={["hover:underline", "text-inherit"]}>
            {head.name}
          </InternalLink>
        </li>
      )
    }
    {tail != null && <li>{tail}</li>}
  </ol>
  <span class:list={["ml-2 h-px flex-grow rounded bg-gray-300"]}></span>
</header>
