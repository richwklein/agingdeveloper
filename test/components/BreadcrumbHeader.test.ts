import BreadcrumbHeader from "@components/BreadcrumbHeader.astro"
import { experimental_AstroContainer as AstroContainer } from "astro/container"
import { beforeAll, describe, expect, test } from "vitest"

type RenderOptions = { tail?: string; className?: string }

describe("BreadcrumbHeader", () => {
  let header: string

  beforeAll(async () => {
    header = await render()
  })

  const render = async ({ tail, className }: RenderOptions = {}) => {
    const container = await AstroContainer.create()
    return await container.renderToString(BreadcrumbHeader, {
      props: {
        head: {
          name: "Head",
          path: "/head",
        },
        tail: tail,
        class: className,
      },
    })
  }
  test("that it contains correct tags", async () => {
    expect(header).toContain("</header")
    expect(header).toContain("</ol>")
    expect(header).toContain("</li>")
  })

  test("that it contains the header without a link", async () => {
    expect(header).toContain("<li>Head</li>")
    expect(header).not.toContain("</a>")
  })

  test("that it contains a link with a tail", async () => {
    const result = await render({ tail: "Tail" })
    expect(result).toContain("Head</a>")
    expect(result).toContain("<li>Tail</li>")
  })

  test("that class is set", async () => {
    const result = await render({ className: "test-class" })
    expect(result).toContain("test-class")
  })
})
