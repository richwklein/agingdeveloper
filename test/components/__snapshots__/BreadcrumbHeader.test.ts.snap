// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BreadcrumbHeader > that it renders correctly. 1`] = `"<!DOCTYPE html><header class="flex w-full items-center" data-astro-source-file="/Users/richard/Development/Projects/agingdeveloper-astro/src/components/BreadcrumbHeader.astro" data-astro-source-loc="26:42"> <span class="mr-2 h-px flex-grow rounded bg-gray-300" data-astro-source-file="/Users/richard/Development/Projects/agingdeveloper-astro/src/components/BreadcrumbHeader.astro" data-astro-source-loc="27:4"></span> <ol class="text-primary inline-flex items-center space-x-0 capitalize md:space-x-3" data-astro-source-file="/Users/richard/Development/Projects/agingdeveloper-astro/src/components/BreadcrumbHeader.astro" data-astro-source-loc="28:87"> <li class="after:mx-1 after:text-gray-300 after:content-['/']" data-astro-source-file="/Users/richard/Development/Projects/agingdeveloper-astro/src/components/BreadcrumbHeader.astro" data-astro-source-loc="33:72"> <!--prettier-ignore --><a data-link="internal" target="_self" href="/test" class="hover:underline text-inherit" data-astro-source-file="/Users/richard/Development/Projects/agingdeveloper-astro/src/components/InternalLink.astro" data-astro-source-loc="21:69">Test</a> </li> <li data-astro-source-file="/Users/richard/Development/Projects/agingdeveloper-astro/src/components/BreadcrumbHeader.astro" data-astro-source-loc="40:26">end</li> </ol> <span class="ml-2 h-px flex-grow rounded bg-gray-300" data-astro-source-file="/Users/richard/Development/Projects/agingdeveloper-astro/src/components/BreadcrumbHeader.astro" data-astro-source-loc="42:4"></span> </header>"`;
